numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -720,7 +721,6 @@ typedef struct tagPyArrayObject {
 } PyArrayObject;
 #endif
 
-#define NPY_SIZEOF_PYARRAYOBJECT (sizeof(PyArrayObject_fields))
 
```

Comment:
What was wrong with this?

Reply:
Nothing as such, but it is a trap, and most code that uses it, such as:
```
struct MySubclass {
    char space_for_ndarray[NPY_SIZEOF_PYARRAYOBJECT];
    int myslot;
}
```
is probably problematic and most likely not be binary compatible with all numpy versions.  So it gives a false sense security?

Reply:
I don't see why, it is defined at compile time just as before. The API is changed anyway, so apps compiled against 1.20 are not guaranteed to be backwards compatible.

Reply:
And code that extends NumPy internals has never been guaranteed to be forward compatible.

Reply:
True, this is only interesting for forward compatibility with possible future size changes.  I would have a faint hope that it may help someone realize that since their module doesn't compile on 1.20+, it has to be modified even if you compile it with an earlier version to run in 1.20+...

I agree, we probably never promised not to modify struct sizes in this way, we have certainly done with many structs (aside from `ndarray` itself).

The first version here, I replaced the macro with `PyArray_Type.tp_basicsize` to make it a runtime constan, which is safe and will still fail if used for static sizes. (It is also probably a pretty useless macro, but...)

Reply:
I suppose the argument is to break the compilation of downstream apps that extend the structure, but if they are compiling against 1.20 they cannot expect to be backwards compatible in any case, and their solution would be the same workaround you use here.

Reply:
> but it cannot do so at compile time.

Could you explain that? `sizeof` is a compiler operator.

> Sizeof is a much used operator in the C or C++. It is a compile time unary operator which can be used to compute the size of its operand.

Reply:
But someone who uses this macro currently will incorrectly believe that their code is forward compatible. They are even avoiding the "deprecated API" to ensure binary compatibility with a potential NumPy release messing with the structs internals.

Reply:
Why not just rename it then? Put an underscore on it or something.

Reply:
At least just comment it out with an explanation of why it is done. The first place people will look for an explanation of trouble will be in the include file.

Reply:
OK, let me make this macro a static assert as Matti posted.  Even if a compiler should not support it the compiler error will point to some text regarding to what is going on...

Reply:
Hmm, although, I am not sure that actually compiles well always (i.e. without super confusing things). The stack-overflow posts seem to a be a bit unsure about it. Some people seem to suggest the weird:
```
static int assert_NPY_SIZEOF_PYARRAYOBJECT_was_removed_because_it_is_misleading_see_1_20_Release_Notes[-1]
```
instead, which is weird, hmmmpf.